lang/src/org/partiql/lang/syntax/PartiQLParser.kt

```diff
@@ -59,26 +61,65 @@ internal class PartiQLParser(
-        val parser = getParser(lexer)
+    /**
+     * To reduce latency costs, the [PartiQLParser] attempts to use [PredictionMode.SLL] and falls back to
+     * [PredictionMode.LL] if a [ParseCancellationException] is thrown by the [BailErrorStrategy]. See [createParserSLL]
+     * and [createParserLL] for more information.
+     */
+    private fun parseQuery(input: String): ParseTree = try {
+        parseUsingSLL(input)
+    } catch (ex: ParseCancellationException) {
+        parseUsingLL(input)
+    }
+
+    internal fun parseUsingSLL(input: String): org.partiql.lang.syntax.antlr.PartiQLParser.StatementContext {
+        val parser = createParserSLL(input)
         return parser.statement()
     }
 
-    private fun getLexer(source: String): Lexer {
-        val inputStream = CharStreams.fromStream(source.byteInputStream(StandardCharsets.UTF_8), StandardCharsets.UTF_8)
-        val lexer = GeneratedLexer(inputStream)
+    internal fun parseUsingLL(input: String): org.partiql.lang.syntax.antlr.PartiQLParser.StatementContext {
+        val parser = createParserLL(input)
+        return parser.statement()
+    }
+
+    private fun createTokenStream(query: String): CommonTokenStream {
+        val inputStream = CharStreams.fromStream(query.byteInputStream(StandardCharsets.UTF_8), StandardCharsets.UTF_8)
         val handler = TokenizeErrorListener(ion)
+        val lexer = GeneratedLexer(inputStream)
         lexer.removeErrorListeners()
         lexer.addErrorListener(handler)
-        return lexer
+        return CommonTokenStream(lexer)
     }
 
-    private fun getParser(lexer: Lexer): GeneratedParser {
-        val tokens = CommonTokenStream(lexer)
-        val parser = GeneratedParser(tokens)
-        val handler = ParseErrorListener(ion)
+    /**
+     * Creates a [GeneratedParser] that uses [PredictionMode.SLL] and the [BailErrorStrategy]. The [GeneratedParser],
+     * upon seeing a syntax error, will throw a [ParseCancellationException] due to the [GeneratedParser.getErrorHandler]
+     * being a [BailErrorStrategy]. The purpose of this is to throw syntax errors as quickly as possible once encountered.
+     * As noted by the [PredictionMode.SLL] documentation, to guarantee results, it is useful to follow-up a failed parse
+     * by parsing with [PredictionMode.LL] -- see [createParserLL] for more information.
+     * See the JavaDocs for [PredictionMode.SLL] and [BailErrorStrategy] for more information.
+     */
+    private fun createParserSLL(query: String): GeneratedParser {
+        val stream = createTokenStream(query)
+        val parser = GeneratedParser(stream)
+        parser.reset()
+        parser.interpreter.predictionMode = PredictionMode.SLL
+        parser.removeErrorListeners()
+        parser.errorHandler = BailErrorStrategy()
+        return parser
+    }
+
+    /**
+     * Creates a [GeneratedParser] that uses [PredictionMode.LL]. This is the fastest prediction mode that guarantees
```

Comment:
I know the point of the statment "This is the fastest prediction mode that guarantees correct parse results." Is supposed to be the "guarantees", but what sticks out is the "fastest".  However, the "LL" method is slower than "SLL".

I'd reword the something like "This method is capable of parsing all valid inputs for a grammar, but is slower than SLL mode.